readme.md
# Colorado Openerr v1.1.0

[Colorado Openerr](https://twitter.com/openerr_co) uses the [Open States API](https://openstates.org/) to tweet about what is happening in Colorado state legislation.

Built by [Ogden Studios](https://ogdenstudios.xyz). The best way to support this project is by giving money to [Open States](https://openstates.org/).

[Read about how and why this bot was made](https://ogdenstudios.xyz/2019/05/21/how-to-build-a-twitter-bot-with-aws-lambda.html).

## Credentials

If you plan to deploy your own version or modification of Openerr, you'll need your own [Open States API key](https://openstates.org/api/register/) and [Twitter API credentials](https://developer.twitter.com/en/docs/basics/getting-started).

## Tests

If you run `npm test`, you'll run the test suite built with [Mocha](https://mochajs.org/). Additionally, if you want to see the twitter output without actually hitting the Twitter API, you can run `npm run local` and instead of tweeting, the script will `console.log()` each tweet status.

This won't be helpful in troubleshooting Twitter API issues, but it might be useful to troubleshoot any issues with the Open States API.

## Deploying to AWS Lambda

This script is meant to be packaged up and deployed to AWS Lambda. If you want to create the `.zip` file for deployment, run `npm run zip`. Current builds are larger than 10mb and should be uploaded to an S3 bucket for lambda deployment.

## Code style

We use the `.editorconfig` and a modified `.eslintrc` from the [Node.js Style Guide](https://github.com/felixge/node-style-guide). The modified `.estlinrc` is only different in that it uses ES6. ESLint has been added to `npm run pretest`, so will fire before testing, or can be run on its own if you like.

You can run `npm run prettier` to run [Prettier](https://prettier.io/) with the correct settings to meet our ESLint config.

## Versioning

We use [semantic versioning](https://semver.org/). Check out `changelog.md` for notes about new versions as they're released.

## License

MIT Licensed. Check `LICENSE` for details.


changelog.md
# 1.1.0

Convert to the [Serverless framework](https://serverless.com/).

# 1.0.0

Add link to blog post and increment version to `1.0.0`.

# 0.2.3

Fix ratelimit handling. If there are over 300 bills, loop gets set to 299 tweets and starts by sending out the ratelimit message.

# 0.2.2

Add a function to construct tweet text and keep it within 280 characters.

# 0.2.1

Add a ratelimit message if we have over 300 bills.

# 0.2.0

Switch over to the `https` module instead of [Axios](https://github.com/axios/axios).

# 0.1.0

MVP working code.


test/test.js
var assert = require('assert');
var openerr = require('../index');
var correctOpenStatesQuery = `
{
    search: bills(first:100, after:"xxx6969xxx", jurisdiction: "Colorado", actionSince: "2019-04-20") {
        edges {
            node {
                id,
                title,
                identifier,
                openstatesUrl,
                actions {
                    description
                    date
                }
            }
        }
        pageInfo {
            hasNextPage
            hasPreviousPage
            endCursor
            startCursor
        }
    }
}
`;
var sampleBillData = {
  node: {
    identifier: 'xxx-xxx-xxx',
    title: 'Correct bill object',
    actions: [
      {
        description: 'Latest action taken',
        date: '2019-04-20T04:20:00',
      },
    ],
    openstatesUrl: 'https://openstates.org',
  },
};
var correctBillObject = {
  identifier: 'xxx-xxx-xxx',
  title: 'Correct bill object',
  latestAction: 'Latest action taken',
  latestActionDate: '2019-04-20',
  openstatesUrl: 'https://openstates.org',
};

var shortBill = {
  identifier: '',
  title: '',
  latestActionDate: '',
  latestAction: '',
  openstatesUrl: '',
};

var truncatedBill = {
  identifier: '',
  title: '',
  latestActionDate: '',
  latestAction: '',
  openstatesUrl:
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec.',
};

var longBill = {
  identifier: 'xxx-xxx-xxx',
  title:
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec.',
  latestActionDate: '',
  latestAction: '',
  openstatesUrl:
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec.',
};

var correctFullTweet = 'Colorado : . On , the following action was taken: .';

var correctTruncatedTweet =
  'Colorado : . On , the following action was taken: .';

var correctDefaultTweet =
  'There was action on Colorado xxx-xxx-xxx, but it was too long to tweet.';

describe('Openerr', function() {
  describe('#createOpenStatesQuery()', function() {
    it('should create the Open States query as expected', function() {
      // Use `.raw` to compare raw strings.
      assert.equal(
        openerr.testCreateOpenStatesQuery('2019-04-20', 'xxx6969xxx'),
        correctOpenStatesQuery.raw
      );
    });
  });
  describe('#createBillObject()', function() {
    it('should create bill objects as expected', function() {
      assert.equal(
        openerr.testCreateBillObject(sampleBillData).identifier,
        correctBillObject.identifier
      );
      assert.equal(
        openerr.testCreateBillObject(sampleBillData).title,
        correctBillObject.title
      );
      assert.equal(
        openerr.testCreateBillObject(sampleBillData).latestAction,
        correctBillObject.latestAction
      );
      assert.equal(
        openerr.testCreateBillObject(sampleBillData).latestActionDate,
        correctBillObject.latestActionDate
      );
      assert.equal(
        openerr.testCreateBillObject(sampleBillData).openstatesUrl,
        correctBillObject.openstatesUrl
      );
    });
  });
  describe('#startTweeting()', function() {
    it('should return false when bills are empty', function() {
      assert.equal(openerr.testStartTweeting([]), false);
    });
    it('should return the bills when they are present', function() {
      assert.equal(openerr.testStartTweeting([correctBillObject]).length, 1);
    });
    it('should return with a too many bills message if there are 600+', function() {
      var tooManyBills = [];
      for (i = 0; i < 700; i++) {
        tooManyBills.push(correctBillObject);
      }
      assert.equal(openerr.testStartTweeting(tooManyBills).length, 700);
    });
  });
  describe('#createTweetText()', function() {
    it('should return full tweet if length < 280', function() {
      assert.equal(
        openerr.testCreateTweetText(shortBill).raw,
        correctFullTweet.raw
      );
    });
    it('should return tweetBody if tweetbody + readMore > 280 && tweetBody < 280', function() {
      assert.equal(
        openerr.testCreateTweetText(truncatedBill).raw,
        correctTruncatedTweet.raw
      );
    });
    it('should return default message if tweetBody > 280', function() {
      assert.equal(
        openerr.testCreateTweetText(longBill).raw,
        correctDefaultTweet.raw
      );
    });
  });
});


openerr.js
// Load dependencies 
const axios = require('axios');
const secrets = require("./secrets.json");
var Twitter = require('twitter');
// Set up constants and variables 
const url = "https://openstates.org/graphql";
var today = new Date();
bills = []
var twitter = new Twitter({
    consumer_key: secrets.api_key,
    consumer_secret: secrets.api_secret_key,
    access_token_key: secrets.access_token,
    access_token_secret: secrets.access_token_secret
});
var date = today.getFullYear() + '-0' + (today.getMonth() + 1) + '-' + today.getDate();
var openStatesQuery = createOpenStatesQuery(date);

function getIt(url, query) {
    axios.get(url, { params: { query: query }, headers: { 'X-API-KEY': secrets.openStatesKey } })
        .then(function (response) {
            var hasNextPage = response.data.data.search.pageInfo.hasNextPage;
            var endCursor = response.data.data.search.pageInfo.endCursor;
            if (hasNextPage) {
                var responseData = response.data.data.search.edges;
                for (i = 0;i < responseData.length;i++) {
                    bill = constructBillObject(responseData[i]);
                    bills.push(bill);
                }
                var newQuery = createOpenStatesQuery(date, endCursor);
                getIt(url, newQuery);
            } else {
                var responseData = response.data.data.search.edges;
                for (i = 0;i < responseData.length;i++) {
                    bill = constructBillObject(responseData[i]);
                    bills.push(bill);
                }
                startTweeting();
            }
        })
        .catch(function (error) {
            console.log(error);
        });
}

getIt(url, openStatesQuery);

// Tweet out the results 
function startTweeting() {
    if (bills.length === 0) {
        tweet("We didn't find any bill activity for yesterday. If you think this is an error, please DM us! For more information about the CO legislature, visit: https://leg.colorado.gov/");
    } else {
        for (i = 0;i < bills.length;i++) {
            var readMore = bills[i].openstatesURL ? "Read more at: " + bills[i].openstatesURL.toString() : '';
            var tweetText = `Colorado ${bills[i].identifier}: ${bills[i].title}. On ${bills[i].latestActionDate}, the following action was taken: ${bills[i].latestAction}. ${readMore}`;
            tweet(tweetText);
        }
    }
}

function tweet(status) {
    twitter.post('statuses/update', { status: status })
        .then(function (tweet) {
            console.log(tweet);
        })
        .catch(function (error) {
            throw error;
        });
}

// Open States Query constructor 
function createOpenStatesQuery(date, cursor = null) {
    var query = `
        { 
            search: bills(first:100, after:"${cursor ? cursor : ''}", jurisdiction: "Colorado", actionSince: "${date}") {
                edges {
                    node {
                        id,
                        title,
                        identifier,
                        openstatesUrl,
                        actions {
                            description
                            date
                        }
                    }
                }
                pageInfo {
                    hasNextPage
                    hasPreviousPage
                    endCursor
                    startCursor
                }
            }
        }
    `;
    return query;
}

// Bill object constructor 
function constructBillObject(data) {
    bill = {
        identifier: data.node.identifier,
        title: data.node.title,
        latestAction: data.node.actions[0].description,
        latestActionDate: data.node.actions[0].date.split("T")[0],
        openstatesURL: data.node.openstatesUrl
    }
    return bill;
}

index.js
var d = new Date();
d.setDate(d.getDate() - 1);
var date = d.toISOString().split('T')[0];
var env = 'test';
var https = require('https');
var openStatesQuery = createOpenStatesQuery(date);
var Twitter = require('twitter');
var twitter = new Twitter({
  consumer_key: process.env.api_key,
  consumer_secret: process.env.api_secret_key,
  access_token_key: process.env.access_token,
  access_token_secret: process.env.access_token_secret,
});
var url = 'openstates.org';

function getIt(query, bills) {
  var options = {
    headers: {'X-API-KEY': process.env.openStatesKey},
    host: url,
    path: `/graphql/?query=${query}`,
  };
  var req = https.get(options, function (res) {
    console.log('STATUS: ' + res.statusCode);
    console.log('HEADERS: ' + JSON.stringify(res.headers));
    var bodyChunks = [];
    res
      .on('data', function (chunk) {
        bodyChunks.push(chunk);
      })
      .on('end', function () {
        var body = Buffer.concat(bodyChunks);
        var parsedBody = JSON.parse(body);
        var hasNextPage = parsedBody.data.search.pageInfo.hasNextPage;
        var endCursor = parsedBody.data.search.pageInfo.endCursor;
        var responseData = parsedBody.data.search.edges;
        for (i = 0;i < responseData.length;i++) {
          bill = createBillObject(responseData[i]);
          bills.push(bill);
        }
        if (hasNextPage) {
          var newQuery = createOpenStatesQuery(date, endCursor);
          getIt(newQuery, bills);
        } else {
          startTweeting(bills);
        }
      });
  });
  req.on('error', function (e) {
    console.log('ERROR: ' + e.message);
  });
}

function startTweeting(bills, testing = false) {
  var limit = 0;
  if (bills.length === 0) {
    console.log('No bills found today');
    return false;
  } else if (bills.length > 300) {
    tweet('There were over 300 bills with activity yesterday, which exceeds our rate limit for sending tweets. Please check https://openstates.org/co/ for more details');
    limit = 299;
  } else {
    limit = bills.length;
  }
  for (i = 0;i < limit;i++) {
    var tweetText = createTweetText(bills[i]);
    if (testing) {
      return bills;
    } else {
      tweet(tweetText);
    }
  }
}

function tweet(status) {
  if (env === 'production') {
    twitter
      .post('statuses/update', {status: status})
      .then(function (tweet) {
        console.log(tweet);
      })
      .catch(function (error) {
        console.log(error);
      });
  } else {
    console.log(status);
  }
}

function createOpenStatesQuery(date, cursor = null) {
  var query = `
        {
            search: bills(first:100, after:"${
    cursor ? cursor : ''
    }", jurisdiction: "Colorado", actionSince: "${date}") {
                edges {
                    node {
                        id,
                        title,
                        identifier,
                        openstatesUrl,
                        actions {
                            description
                            date
                        }
                    }
                }
                pageInfo {
                    hasNextPage
                    hasPreviousPage
                    endCursor
                    startCursor
                }
            }
        }
    `;
  return encodeURIComponent(query);
}

function createBillObject(data) {
  bill = {
    identifier: data.node.identifier,
    title: data.node.title,
    latestAction: data.node.actions[0].description,
    latestActionDate: data.node.actions[0].date.split('T')[0],
    openstatesUrl: data.node.openstatesUrl,
  };
  return bill;
}

function createTweetText(bill) {
  var tweetText = '';
  var tweetBody = `Colorado ${bill.identifier}: ${bill.title}. On ${
    bill.latestActionDate
    }, the following action was taken: ${bill.latestAction}. `;
  var readMore = bill.openstatesUrl
    ? 'Read more at: ' + bill.openstatesUrl.toString()
    : '';
    if ((tweetBody + readMore).length <= 280) {
      tweetText = tweetBody + readMore;
    } else if (tweetBody.length <= 280) {
      tweetText = tweetBody;
    } else {
      tweetText = `There was action on Colorado ${bill.identifier}, but it was too long to tweet.`;
    }
  return tweetText;
}

exports.handler = function (event, context, callback) {
  console.log(event);
  console.log(context);
  env = 'production';
  getIt(openStatesQuery, []);
  callback(null, 'Success from lambda');
};

exports.testCreateOpenStatesQuery = function (date, cursor) {
  return createOpenStatesQuery(date, cursor).raw;
};

exports.testCreateBillObject = function (data) {
  return createBillObject(data);
};

exports.testStartTweeting = function (bills) {
  return startTweeting(bills, true);
};

exports.testCreateTweetText = function(bill) {
  return createTweetText(bill);
};

if (process.argv[2] === 'local') {
  getIt(openStatesQuery, []);
}
